packages/react-hooks/use-local-storage/useLocalStorage.ts

```diff
@@ -0,0 +1,48 @@
+import { Dispatch, SetStateAction, useCallback, useState } from "react";
+
+const storeValue = <T>(key: string, value: T): void => {
+  if (!window) return;
+
+  const valueSerialized = JSON.stringify(value);
+  localStorage.setItem(key, valueSerialized);
+};
+
+const getInitialStoredValue = <T>(key: string, initialValue: T): T => {
+  if (!window) return initialValue;
+
+  const storedValue = localStorage.getItem(key);
+  if (storedValue !== null) return JSON.parse(storedValue);
+  else {
+    storeValue(key, initialValue);
+    return initialValue;
+  }
+};
+
+/**
+ * Persist the state with local storage so that it remains after a page refresh. This can be useful
+ * for a dark theme or to record session information. This hook is used in the same way as useState
+ * except that you must pass the storage key in the 1st parameter. If the window object is not
+ * present (as in SSR), useLocalStorage() will return the default value.
+ *
+ * @param key - Key to use in local storage to store value
+ */
+const useLocalStorage = <T>(key: string, initialValue: T): [T, Dispatch<SetStateAction<T>>] | [T] => {
+  const [state, setState] = useState(getInitialStoredValue(key, initialValue));
+
+  const storeState: Dispatch<SetStateAction<T>> = useCallback(
+    (value) => {
+      let newValue = value;
+      if (typeof value === "function") {
+        newValue = (value as ((prevState: T) => T))(state);
+      }
+      setState(newValue);
+      storeValue(key, newValue);
+    },
+    [key, state]
+  );
+
+  if (!window) return [initialValue as T];
```

Comment:
we need to respect still return the setter in order to make it act as `useState` by default